From ef324ff7afb28cd2a807ed083d85efdc6a4b6970 Mon Sep 17 00:00:00 2001
From: "Lu.Jiang" <lu.jiang@windriver.com>
Date: Tue, 16 Sep 2014 11:10:06 +0800
Subject: [PATCH 0942/1089] iommu/fsl: Work around erratum A-007907

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

Erratum A-007907 can cause a core hang under certain circumstances.
Part of the workaround involves not stashing to L1 Cache.  On affected
chips, stash to L2 when L1 is requested.

Signed-off-by: Scott Wood <scottwood@freescale.com>
Change-Id: Ie8b16bf84831aae2b97897e4da91f91e8bdd9868
Reviewed-on: http://git.am.freescale.net:8181/11208
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Signed-off-by: Lu.Jiang <lu.jiang@windriver.com>
---
 drivers/iommu/fsl_pamu.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/drivers/iommu/fsl_pamu.c b/drivers/iommu/fsl_pamu.c
index 1481029..78f10ea 100644
--- a/drivers/iommu/fsl_pamu.c
+++ b/drivers/iommu/fsl_pamu.c
@@ -34,6 +34,7 @@
 #include <asm/fsl_guts.h>
 #include <asm/fsl_kibo.h>
 #include <linux/syscore_ops.h>
+#include <asm/mpc85xx.h>
 
 #include "fsl_pamu.h"
 
@@ -662,6 +663,26 @@ static u32 get_dsp_l2_stash_id(u32 vcpu)
 	return ~(u32)0;
 }
 
+static bool has_erratum_a007907(void)
+{
+	u32 svr = mfspr(SPRN_SVR);
+
+	switch (SVR_SOC_VER(svr)) {
+	case SVR_B4860:
+	case SVR_B4420:
+	case SVR_T4240:
+	case SVR_T4160:
+		return SVR_REV(svr) <= 0x20;
+
+	case SVR_T2080:
+	case SVR_T2081:
+		return SVR_REV(svr) == 0x10;
+
+	default:
+		return false;
+	};
+}
+
 /**
  * get_stash_id - Returns stash destination id corresponding to a
  *                cache type and vcpu.
@@ -679,6 +700,10 @@ u32 get_stash_id(u32 stash_dest_hint, u32 vcpu)
 	int len, found = 0;
 	int i;
 
+	if (stash_dest_hint == PAMU_ATTR_CACHE_L1 &&
+	    has_erratum_a007907())
+		stash_dest_hint = PAMU_ATTR_CACHE_L2;
+
 	/* check for DSP L2 cache */
 	if (stash_dest_hint == PAMU_ATTR_CACHE_DSP_L2) {
 		return get_dsp_l2_stash_id(vcpu);
-- 
2.0.2

